### ISSUE TITLE
use after free issue in ast.h 511

### ISSUE THREAD

**Issue by 5hadowblad3** [Z3Prover/z3#3247]
Hi, there.

There is a use after free issue that causes segmentation fault using z3.

To reproduce this issue, simply run
z3 poc.smt2
[poc2.smt2.zip](https://github.com/Z3Prover/z3/files/4322773/poc2.smt2.zip)

This is the minimized version of the formula.
~~~~
(set-logic QF_BV)
(declare-const bv_21-0 (_ BitVec 21))
(assert (xor true true true true true true true (= (bvashr (_ bv0 29) (concat (_ bv204 8) bv_21-0)) (_ bv0 29) (_ bv0 29)) true true true))
(check-sat)
(exit)
~~~~

OS: Ubuntu 16.06
commit 7bcd345

~~~~
==107318==ERROR: AddressSanitizer: heap-use-after-free on address 0x6070004d0350 at pc 0x000000b98ee6 bp 0x7ffeee117770 sp 0x7ffeee117760
READ of size 8 at 0x6070004d0350 thread T0
    #0 0xb98ee5 in ast::reset_mark2(void*) ../src/ast/ast.h:511
    #1 0xbbacea in ast_fast_mark<2u>::reset_mark(ast*) ../src/ast/ast.h:2484
    #2 0xbb57e6 in ast_fast_mark<2u>::reset() ../src/ast/ast.h:2502
    #3 0xbb2213 in ast_fast_mark<2u>::~ast_fast_mark() ../src/ast/ast.h:2476
    #4 0x1ace4ff in bool_rewriter::mk_nflat_or_core(unsigned int, expr* const*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bool_rewriter.cpp:180
    #5 0x656d0b in bool_rewriter::mk_or_core(unsigned int, expr* const*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bool_rewriter.h:125
    #6 0x656e27 in bool_rewriter::mk_or(unsigned int, expr* const*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bool_rewriter.h:145
    #7 0x9e5066 in bool_rewriter::mk_or(expr*, expr*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bool_rewriter.h:154
    #8 0x1581ba1 in blaster_cfg::mk_or(expr*, expr*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bit_blaster/bit_blaster_rewriter.cpp:46
    #9 0x15917f1 in bit_blaster_tpl<blaster_cfg>::mk_or(expr*, expr*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bit_blaster/bit_blaster_tpl.h:66
    #10 0x158a967 in bit_blaster_tpl<blaster_cfg>::mk_ashr(unsigned int, expr* const*, expr* const*, ref_vector<expr, ast_manager>&) ../src/ast/rewriter/bit_blaster/bit_blaster_tpl_def.h:1042
    #11 0x1583bfe in blaster_rewriter_cfg::reduce_ashr(expr*, expr*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bit_blaster/bit_blaster_rewriter.cpp:259
    #12 0x158790e in blaster_rewriter_cfg::reduce_app(func_decl*, unsigned int, expr* const*, obj_ref<expr, ast_manager>&, obj_ref<app, ast_manager>&) ../src/ast/rewriter/bit_blaster/bit_blaster_rewriter.cpp:512
    #13 0x15a0ae9 in void rewriter_tpl<blaster_rewriter_cfg>::process_app<false>(app*, rewriter_core::frame&) ../src/ast/rewriter/rewriter_def.h:291
    #14 0x159a373 in void rewriter_tpl<blaster_rewriter_cfg>::resume_core<false>(obj_ref<expr, ast_manager>&, obj_ref<app, ast_manager>&) ../src/ast/rewriter/rewriter_def.h:742
    #15 0x1596e98 in void rewriter_tpl<blaster_rewriter_cfg>::main_loop<false>(expr*, obj_ref<expr, ast_manager>&, obj_ref<app, ast_manager>&) ../src/ast/rewriter/rewriter_def.h:703
    #16 0x1591268 in rewriter_tpl<blaster_rewriter_cfg>::operator()(expr*, obj_ref<expr, ast_manager>&, obj_ref<app, ast_manager>&) ../src/ast/rewriter/rewriter_def.h:773
    #17 0x15814fc in bit_blaster_rewriter::operator()(expr*, obj_ref<expr, ast_manager>&, obj_ref<app, ast_manager>&) ../src/ast/rewriter/bit_blaster/bit_blaster_rewriter.cpp:722
    #18 0xca9199 in bit_blaster_tactic::imp::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/bv/bit_blaster_tactic.cpp:74
    #19 0xca9f43 in bit_blaster_tactic::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/bv/bit_blaster_tactic.cpp:139
    #20 0x19f141f in cleanup_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/tactical.cpp:909
    #21 0x19ea1ec in and_then_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/tactical.cpp:111
    #22 0x19ea33b in and_then_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/tactical.cpp:120
    #23 0x19ea33b in and_then_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/tactical.cpp:120
    #24 0x19ea33b in and_then_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/tactical.cpp:120
    #25 0x19ea33b in and_then_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/tactical.cpp:120
    #26 0x19ea33b in and_then_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/tactical.cpp:120
    #27 0xaa6f71 in inc_sat_solver::internalize_goal(ref<goal>&, obj_map<expr, sat::literal>&, bool) ../src/sat/sat_solver/inc_sat_solver.cpp:592
    #28 0xaa9ec9 in inc_sat_solver::internalize_formulas() (/home/heqing/dependence/z3/build/z3+0xaa9ec9)
    #29 0xaa214b in inc_sat_solver::check_sat_core(unsigned int, expr* const*) ../src/sat/sat_solver/inc_sat_solver.cpp:190
    #30 0x1978222 in combined_solver::check_sat_core(unsigned int, expr* const*) ../src/solver/combined_solver.cpp:219
    #31 0x197575f in solver::check_sat(unsigned int, expr* const*) ../src/solver/solver.cpp:330
    #32 0x192ec2a in cmd_context::check_sat(unsigned int, expr* const*) ../src/cmd_context/cmd_context.cpp:1581
    #33 0x18bebfb in smt2::parser::parse_check_sat() ../src/parsers/smt2/smt2parser.cpp:2596
    #34 0x18c2b97 in smt2::parser::parse_cmd() ../src/parsers/smt2/smt2parser.cpp:2938
    #35 0x18c42a9 in smt2::parser::operator()() ../src/parsers/smt2/smt2parser.cpp:3130
    #36 0x18a328e in parse_smt2_commands(cmd_context&, std::istream&, bool, params_ref const&, char const*) ../src/parsers/smt2/smt2parser.cpp:3179
    #37 0x43c05c in read_smtlib2_commands(char const*) ../src/shell/smtlib_frontend.cpp:89
    #38 0x453e68 in main ../src/shell/main.cpp:352
    #39 0x7f55882df82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #40 0x414808 in _start (/home/heqing/dependence/z3/build/z3+0x414808)

0x6070004d0350 is located 32 bytes inside of 76-byte region [0x6070004d0330,0x6070004d037c)
freed by thread T0 here:
    #0 0x7f55891df2ca in __interceptor_free (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x982ca)
    #1 0x255ad17 in memory::deallocate(void*) ../src/util/memory_manager.cpp:260
    #2 0x258ad02 in small_object_allocator::deallocate(unsigned long, void*) ../src/util/small_object_allocator.cpp:77
    #3 0x21f560c in ast_manager::deallocate_node(ast*, unsigned int) ../src/ast/ast.h:1733
    #4 0x21e313e in ast_manager::delete_node(ast*) ../src/ast/ast.cpp:1997
    #5 0x43f27c in ast_manager::dec_ref(ast*) ../src/ast/ast.h:1692
    #6 0x44201b in obj_ref<expr, ast_manager>::dec_ref() ../src/util/obj_ref.h:33
    #7 0x4717bc in obj_ref<expr, ast_manager>::operator=(expr*) ../src/util/obj_ref.h:82
    #8 0x1ace048 in bool_rewriter::mk_nflat_or_core(unsigned int, expr* const*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bool_rewriter.cpp:185
    #9 0x656d0b in bool_rewriter::mk_or_core(unsigned int, expr* const*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bool_rewriter.h:125
    #10 0x656e27 in bool_rewriter::mk_or(unsigned int, expr* const*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bool_rewriter.h:145
    #11 0x9e5066 in bool_rewriter::mk_or(expr*, expr*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bool_rewriter.h:154
    #12 0x1581ba1 in blaster_cfg::mk_or(expr*, expr*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bit_blaster/bit_blaster_rewriter.cpp:46
    #13 0x15917f1 in bit_blaster_tpl<blaster_cfg>::mk_or(expr*, expr*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bit_blaster/bit_blaster_tpl.h:66
    #14 0x158a967 in bit_blaster_tpl<blaster_cfg>::mk_ashr(unsigned int, expr* const*, expr* const*, ref_vector<expr, ast_manager>&) ../src/ast/rewriter/bit_blaster/bit_blaster_tpl_def.h:1042
    #15 0x1583bfe in blaster_rewriter_cfg::reduce_ashr(expr*, expr*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bit_blaster/bit_blaster_rewriter.cpp:259
    #16 0x158790e in blaster_rewriter_cfg::reduce_app(func_decl*, unsigned int, expr* const*, obj_ref<expr, ast_manager>&, obj_ref<app, ast_manager>&) ../src/ast/rewriter/bit_blaster/bit_blaster_rewriter.cpp:512
    #17 0x15a0ae9 in void rewriter_tpl<blaster_rewriter_cfg>::process_app<false>(app*, rewriter_core::frame&) ../src/ast/rewriter/rewriter_def.h:291
    #18 0x159a373 in void rewriter_tpl<blaster_rewriter_cfg>::resume_core<false>(obj_ref<expr, ast_manager>&, obj_ref<app, ast_manager>&) ../src/ast/rewriter/rewriter_def.h:742
    #19 0x1596e98 in void rewriter_tpl<blaster_rewriter_cfg>::main_loop<false>(expr*, obj_ref<expr, ast_manager>&, obj_ref<app, ast_manager>&) ../src/ast/rewriter/rewriter_def.h:703
    #20 0x1591268 in rewriter_tpl<blaster_rewriter_cfg>::operator()(expr*, obj_ref<expr, ast_manager>&, obj_ref<app, ast_manager>&) ../src/ast/rewriter/rewriter_def.h:773
    #21 0x15814fc in bit_blaster_rewriter::operator()(expr*, obj_ref<expr, ast_manager>&, obj_ref<app, ast_manager>&) ../src/ast/rewriter/bit_blaster/bit_blaster_rewriter.cpp:722
    #22 0xca9199 in bit_blaster_tactic::imp::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/bv/bit_blaster_tactic.cpp:74
    #23 0xca9f43 in bit_blaster_tactic::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/bv/bit_blaster_tactic.cpp:139
    #24 0x19f141f in cleanup_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/tactical.cpp:909
    #25 0x19ea1ec in and_then_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/tactical.cpp:111
    #26 0x19ea33b in and_then_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/tactical.cpp:120
    #27 0x19ea33b in and_then_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/tactical.cpp:120
    #28 0x19ea33b in and_then_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/tactical.cpp:120
    #29 0x19ea33b in and_then_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/tactical.cpp:120

previously allocated by thread T0 here:
    #0 0x7f55891df602 in malloc (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x98602)
    #1 0x255ad63 in memory::allocate(unsigned long) ../src/util/memory_manager.cpp:268
    #2 0x258ad33 in small_object_allocator::allocate(unsigned long) ../src/util/small_object_allocator.cpp:102
    #3 0x21f55dc in ast_manager::allocate_node(unsigned int) ../src/ast/ast.h:1729
    #4 0x21e54fd in ast_manager::mk_app_core(func_decl*, unsigned int, expr* const*) ../src/ast/ast.cpp:2235
    #5 0x21e69e8 in ast_manager::mk_app(func_decl*, unsigned int, expr* const*) ../src/ast/ast.cpp:2351
    #6 0x21e33d3 in ast_manager::mk_app(int, int, unsigned int, parameter const*, unsigned int, expr* const*, sort*) ../src/ast/ast.cpp:2028
    #7 0x21e341e in ast_manager::mk_app(int, int, unsigned int, expr* const*) ../src/ast/ast.cpp:2033
    #8 0x5b9148 in ast_manager::mk_or(unsigned int, expr* const*) ../src/ast/ast.h:2170
    #9 0x1ace4cd in bool_rewriter::mk_nflat_or_core(unsigned int, expr* const*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bool_rewriter.cpp:239
    #10 0x656d0b in bool_rewriter::mk_or_core(unsigned int, expr* const*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bool_rewriter.h:125
    #11 0x656e27 in bool_rewriter::mk_or(unsigned int, expr* const*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bool_rewriter.h:145
    #12 0x9e5066 in bool_rewriter::mk_or(expr*, expr*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bool_rewriter.h:154
    #13 0x1581ba1 in blaster_cfg::mk_or(expr*, expr*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bit_blaster/bit_blaster_rewriter.cpp:46
    #14 0x15917f1 in bit_blaster_tpl<blaster_cfg>::mk_or(expr*, expr*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bit_blaster/bit_blaster_tpl.h:66
    #15 0x158a967 in bit_blaster_tpl<blaster_cfg>::mk_ashr(unsigned int, expr* const*, expr* const*, ref_vector<expr, ast_manager>&) ../src/ast/rewriter/bit_blaster/bit_blaster_tpl_def.h:1042
    #16 0x1583bfe in blaster_rewriter_cfg::reduce_ashr(expr*, expr*, obj_ref<expr, ast_manager>&) ../src/ast/rewriter/bit_blaster/bit_blaster_rewriter.cpp:259
    #17 0x158790e in blaster_rewriter_cfg::reduce_app(func_decl*, unsigned int, expr* const*, obj_ref<expr, ast_manager>&, obj_ref<app, ast_manager>&) ../src/ast/rewriter/bit_blaster/bit_blaster_rewriter.cpp:512
    #18 0x15a0ae9 in void rewriter_tpl<blaster_rewriter_cfg>::process_app<false>(app*, rewriter_core::frame&) ../src/ast/rewriter/rewriter_def.h:291
    #19 0x159a373 in void rewriter_tpl<blaster_rewriter_cfg>::resume_core<false>(obj_ref<expr, ast_manager>&, obj_ref<app, ast_manager>&) ../src/ast/rewriter/rewriter_def.h:742
    #20 0x1596e98 in void rewriter_tpl<blaster_rewriter_cfg>::main_loop<false>(expr*, obj_ref<expr, ast_manager>&, obj_ref<app, ast_manager>&) ../src/ast/rewriter/rewriter_def.h:703
    #21 0x1591268 in rewriter_tpl<blaster_rewriter_cfg>::operator()(expr*, obj_ref<expr, ast_manager>&, obj_ref<app, ast_manager>&) ../src/ast/rewriter/rewriter_def.h:773
    #22 0x15814fc in bit_blaster_rewriter::operator()(expr*, obj_ref<expr, ast_manager>&, obj_ref<app, ast_manager>&) ../src/ast/rewriter/bit_blaster/bit_blaster_rewriter.cpp:722
    #23 0xca9199 in bit_blaster_tactic::imp::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/bv/bit_blaster_tactic.cpp:74
    #24 0xca9f43 in bit_blaster_tactic::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/bv/bit_blaster_tactic.cpp:139
    #25 0x19f141f in cleanup_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/tactical.cpp:909
    #26 0x19ea1ec in and_then_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/tactical.cpp:111
    #27 0x19ea33b in and_then_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/tactical.cpp:120
    #28 0x19ea33b in and_then_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/tactical.cpp:120
    #29 0x19ea33b in and_then_tactical::operator()(ref<goal> const&, sref_buffer<goal, 16u>&) ../src/tactic/tactical.cpp:120

SUMMARY: AddressSanitizer: heap-use-after-free ../src/ast/ast.h:511 ast::reset_mark2(void*)
Shadow bytes around the buggy address:
  0x0c0e80092010: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e80092020: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e80092030: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e80092040: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0e80092050: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c0e80092060: fa fa fa fa fa fa fd fd fd fd[fd]fd fd fd fd fd
  0x0c0e80092070: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fa fa
  0x0c0e80092080: fa fa fd fd fd fd fd fd fd fd fd fd fa fa fa fa
  0x0c0e80092090: fd fd fd fd fd fd fd fd fd fd fa fa fa fa fd fd
  0x0c0e800920a0: fd fd fd fd fd fd fd fd fa fa fa fa fd fd fd fd
  0x0c0e800920b0: fd fd fd fd fd fd fa fa fa fa fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
==107318==ABORTING
~~~~

**Comment by NikolajBjorner** [comment]
fixed by #3248

**Comment by rainoftime** [comment]
The issue still exists in https://github.com/Z3Prover/z3/issues/3247
[ast511.txt](https://github.com/Z3Prover/z3/files/4332307/ast511.txt)
